lib/pac55xx/gpio.c

```diff
@@ -0,0 +1,164 @@
+ */
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * This library is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <libopencm3/pac55xx/gpio.h>
+
+static uint32_t get_ccs_port_base(uint32_t gpioport) {
+	switch (gpioport) {
+		case GPIOA:
+			return CCS_PORTA;
+		case GPIOB:
+			return CCS_PORTB;
+		case GPIOC:
+			return CCS_PORTC;
+		case GPIOD:
+			return CCS_PORTD;
+		case GPIOE:
+			return CCS_PORTE;
+		case GPIOF:
+			return CCS_PORTF;
+		case GPIOG:
+			return CCS_PORTG;
+		default:
+			return 0U;
+	}
+}
+
+void gpio_mode_setup(uint32_t gpioport, gpio_mode_t mode,
+		     ccs_pull_updown_t pull_up_down, uint16_t gpios) {
+	/* Read the current value of the register. */
+	uint32_t reg  = GPIO_MODER(gpioport);
+	uint32_t port = get_ccs_port_base(gpioport);
+
+	/* Loop through only set bits, utilize built-ins for optimized assembly. */
+	int ffs = __builtin_ffs(gpios);
+	while (ffs) {
+		const int pin = ffs - 1;
+		const int bit = (1 << pin);
+
+		/* Update the cached mode value by clearing then setting values. */
+		reg &= ~GPIO_MODER_MASK_PIN(pin);
+		reg |= GPIO_MODER_MODE(pin, mode);
+
+		/* Set the pinmux configurations for the pull-up / pull-down. */
+		if (pull_up_down == CCS_IO_PULL_UP) {
+			CCS_PDENR(port) &= ~bit;
+			CCS_PUENR(port) |= bit;
+		} else if (pull_up_down == CCS_IO_PULL_DOWN) {
+			CCS_PUENR(port) &= ~bit;
+			CCS_PDENR(port) |= bit;
+		} else {
+			CCS_PDENR(port) &= ~bit;
+			CCS_PUENR(port) &= ~bit;
+		}
+		gpios ^= bit; /* Clear the bit we just serviced. */
+		ffs = __builtin_ffs(gpios);
+	}
+	GPIO_MODER(gpioport) = reg;
+}
+
+void gpio_set_outmask(uint32_t gpioport, bool enable, uint16_t gpios) {
+	uint32_t reg = GPIO_OUTMASKR(gpioport);
+	if (enable) {
+		reg |= gpios;
+	} else {
+		reg &= ~gpios;
+	}
+	GPIO_OUTMASKR(gpioport) = reg;
+}
+
+void gpio_set(uint32_t gpioport, uint16_t gpios) {
+	GPIO_DOSETR(gpioport) = gpios;
+}
+
+void gpio_clear(uint32_t gpioport, uint16_t gpios) {
+	GPIO_DOCLEARR(gpioport) = gpios;
+}
+
+uint16_t gpio_get(uint32_t gpioport, uint16_t gpios) {
+	return GPIO_INR(gpioport) & gpios;
+}
+
+void gpio_set_af(uint32_t gpioport, ccs_muxsel_func_t muxsel, uint16_t gpios) {
+	uint32_t port = get_ccs_port_base(gpioport);
+
+	/* Update each of the pin configs. */
+	uint32_t reg = CCS_MUXSELR(port);
+	int ffs      = __builtin_ffs(gpios);
```

Comment:
we've never done this elsewhere, I dont' see any reason to start, I'm guessing this is from a "suboptimal" choice of options when you clang-formatted the whole thing?

Reply:
Adjacent assignments are aligned for readability. I can remove this option.